File: C/15_DP1/BOJ_12_11054_가장긴바이토닉부분수열.c
#include <stdio.h>
#define math_max(a, b) a > b ? a : b
main() {
	int n;
	scanf("%d", &n);
	int* arr = (int*)malloc(sizeof(int) * n);
	int* dp = (int*)malloc(sizeof(int) * n);
	int* dp2 = (int*)malloc(sizeof(int) * n);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
		dp[i] = dp2[i] = 1;
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < i; j++) {
			if (arr[i] > arr[j] && dp[i] < dp[j] + 1) {
				dp[i] = dp[j] + 1;
			}
		}
	}
	for (int i = n - 1; i >= 0; i--) {
		for (int j = n - 1; j > i; j--) {
			if (arr[i] > arr[j] && dp2[i] < dp2[j] + 1) {
				dp2[i] = dp2[j] + 1;
			}
		}
	}
	int max = 1;
	for (int i = 0; i < n; i++) {
		max = math_max(max, dp[i] + dp2[i]);
	}
	printf("%d", max - 1);
}

File: JAVA/15_DP1/BOJ_12_11054_가장긴바이토닉부분수열.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int[] arr;
    static Integer[] dp;
    static Integer[] dp2;

    public static int LIS(int n) {
        if (dp[n] == null) {
            dp[n] = 1;
            for (int i = n - 1; i >= 0; i--) {
                if (arr[i] < arr[n]) {
                    dp[n] = Math.max(dp[n], LIS(i) + 1);
                }
            }
        }
        return dp[n];
    }

    public static int LDS(int n) {
        if (dp2[n] == null) {
            dp2[n] = 1;
            for (int i = n + 1; i < dp2.length; i++) {
                if (arr[i] < arr[n]) {
                    dp2[n] = Math.max(dp2[n], LDS(i) + 1);
                }
            }
        }
        return dp2[n];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        arr = new int[n];
        dp = new Integer[n];
        dp2 = new Integer[n];
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        for (int i = 0; i < n; i++) {
            LIS(i);
            LDS(i);
        }
        int max = 1;
        for (int i = 0; i < n; i++) {
            max = Math.max(max, dp[i] + dp2[i]);
        }
        System.out.println(max - 1);
    }
}